src/shared/openssl-util.c

```diff
@@ -237,6 +163,90 @@ int pubkey_fingerprint(EVP_PKEY *pk, const EVP_MD *md, void **ret, size_t *ret_s
 
         *ret = TAKE_PTR(h);
         *ret_size = msz;
+        return 0;
+}
+
+static int pkey_generate_ec_key(int nid, EVP_PKEY **ret_ppkey) {
+
+        _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *pctx = NULL;
+        _cleanup_(EVP_PKEY_CTX_freep) EVP_PKEY_CTX *kctx = NULL;
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *params = NULL;
+        _cleanup_(EVP_PKEY_freep) EVP_PKEY *ppkey = NULL;
+        int r;
+
+        assert(ret_ppkey);
+        assert_se(nid != NID_undef);
+
+        pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+        if (pctx)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to allocate pkey context");
+
+        r = EVP_PKEY_paramgen_init(pctx);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to initialze pkey parameters context");
+
+        r = EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, nid);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to set pkey context ec curve nid");
+
+        r = EVP_PKEY_paramgen(pctx, &params);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to generate pkey parameters");
+
+        kctx = EVP_PKEY_CTX_new(params, NULL);
+        if (!kctx)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to allocate pkey context");
+
+        r = EVP_PKEY_keygen_init(kctx);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to initialize pkey keygen context");
+
+        r = EVP_PKEY_keygen(kctx, &ppkey);
+        if (r != 1)
+                return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to generate pkey");
+
+        *ret_ppkey = TAKE_PTR(ppkey);
+        return 0;
+}
+
+static int pkey_ecdh_derive_shared_secret(
+                EVP_PKEY *pkey,
+                EVP_PKEY *peer_key,
+                uint8_t *ret_shared_secret,
```

Comment:
should probably be `void*`, it's a some arbitrary "blob" for us, not a byte array